addons/product/models/product.py

```diff
@@ -0,0 +1,865 @@
+        ''' Store the standard price change in order to be able to retrieve the cost of a product for a given date'''
+        self.env['product.price.history'].create({
+            'product_id': self.id,
+            'cost': value,
+            'company_id': self.env.context.get('force_company') or self.env.user.company_id.id
+            })
+
+    @api.constrains('attribute_value_ids')
+    def _check_attribute_value_ids(self):
+        for product in self:
+            attributes = set()
+            for value in product.attribute_value_ids:
+                if value.attribute_id in attributes:
+                    raise ValidationError(_("Error! It is not allowed to choose more than one value for a given attribute."))
+                else:
+                    attributes.add(value.attribute_id)
+        return True
+
+    @api.model
+    def create(self, vals):
+        product_product = super(ProductProduct, self.with_context(create_product_product=True)).create(vals)
+        product_product._set_standard_price(vals.get('standard_price', 0.0))
+        return product_product
+
+    @api.multi
+    def write(self, vals):
+        ''' Store the standard price change in order to be able to retrieve the cost of a product for a given date'''
+        res = super(ProductProduct, self).write(vals)
+        if 'standard_price' in vals:
+            for product in self:
+                product._set_standard_price(vals['standard_price'])
+        return res
+
+    @api.multi
+    def copy(self, default=None):
+        default = default or {}
+
+        if self.env.context.get('variant'):
+            # if we copy a variant or create one, we keep the same template
+            default['product_tmpl_id'] = self.product_tmpl_id.id
+        elif 'name' not in default:
+            default['name'] = _("%s (copy)") % (self.name,)
+
+        return super(ProductProduct, self).copy(default=default)
+
+    @api.model
+    def search(self, args, offset=0, limit=None, order=None, count=False):
+        if self.env.context.get('search_default_categ_id'):
+            args.append((('categ_id', 'child_of', self.env.context['search_default_categ_id'])))
+        return super(ProductProduct, self).search(args, offset=offset, limit=limit, order=order, count=count)
+
+    @api.multi
+    def unlink(self):
+        unlink_ids = []
+        ProductTemplate = self.env['product.template']
+        for product in self:
+            # Check if product still exists, in case it has been unlinked by unlinking its template
+            if not product.exists():
+                continue
+            # Check if the product is last product of this template
+            other_product_ids = self.search([('product_tmpl_id', '=', product.product_tmpl_id.id), ('id', '!=', product.id)])
+            if not other_product_ids:
+                ProductTemplate += product.product_tmpl_id
+            unlink_ids.append(product.id)
+        products = self.browse(unlink_ids)
```

Comment:
Why you append 'id' instead of recordset(s). Use set operation to concatenate recordset(s) .